File: java/Ice/oauth/InterceptorI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import java.util.concurrent.CompletionStage;
import com.zeroc.demos.Ice.oauth.Demo.*;

public class InterceptorI extends com.zeroc.Ice.DispatchInterceptor
{
    private final java.util.List<String> securedOperations;
    private final com.zeroc.Ice.Object servant;
    private final OAuthProvider provider;

    public InterceptorI(com.zeroc.Ice.Object servant, OAuthProvider provider)
    {
        this.servant = servant;
        this.provider = provider;

        securedOperations = new java.util.ArrayList<String>();
        //
        // We only require authorization for the 'setTemp' operation.
        //
        securedOperations.add("setTemp");
    }

    @Override
    public CompletionStage<com.zeroc.Ice.OutputStream> dispatch(com.zeroc.Ice.Request request)
        throws com.zeroc.Ice.UserException
    {
        com.zeroc.Ice.Current current = request.getCurrent();
        //TODO
        current.encoding = new com.zeroc.Ice.EncodingVersion((byte)1, (byte)1);
        //
        // Check if the operation requires authorization to invoke.
        //
        if(securedOperations.contains(current.operation))
        {
            //
            // Validate the client's access token before dispatching to the servant.
            //
            if(!provider.checkToken(current.ctx.get("accessToken")))
            {
                throw new AuthorizationException();
            }
        }
        return servant.ice_dispatch(request);
    }
}

File: cpp11/Ice/interceptor/InterceptorI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <InterceptorI.h>

using namespace std;

InterceptorI::InterceptorI(shared_ptr<Ice::Object> servant, shared_ptr<AuthenticatorI> authenticator,
                           unordered_set<string> securedOperations) :
    _servant(move(servant)), _authenticator(move(authenticator)), _securedOperations(move(securedOperations))
{
}

bool
InterceptorI::dispatch(Ice::Request& request)
{
    auto current = request.getCurrent();
    //
    // Check if the operation requires authorization to invoke.
    //
    if(_securedOperations.find(current.operation) != _securedOperations.end())
    {
        //
        // Validate the client's access token before dispatching to the servant.
        // 'validateToken' throws an exception if the token is invalid or expired.
        //
        auto tokenValue = current.ctx.find("accessToken");
        if(tokenValue != current.ctx.end())
        {
            _authenticator->validateToken(tokenValue->second);
        }
        else
        {
            //
            // If the client didn't include an accessToken, throw an exception.
            //
            throw Demo::AuthorizationException();
        }
    }
    return _servant->ice_dispatch(request);
}
